pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,656 @@
+			return searchForReadyWorkspaceInLocalInformers(h, logicalClusterName, isHome, userName)
+		},
+		tryToCreate: func(ctx context.Context, userName string, logicalClusterName logicalcluster.Name, workspaceType tenancyv1alpha1.ClusterWorkspaceTypeName) (retryAfterSeconds int, createError error) {
+			return tryToCreate(h, ctx, userName, logicalClusterName, workspaceType)
+		},
+	}
+
+	return h
+}
+
+func (h *homeWorkspaceHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	if !h.localInformers.synced() {
+		h.apiHandler.ServeHTTP(rw, req)
+		return
+	}
+
+	ctx := req.Context()
+	lcluster, err := request.ValidClusterFrom(ctx)
+	if err != nil {
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	user, ok := request.UserFrom(ctx)
+	if !ok {
+		err := errors.New("No user in HomeWorkspaces filter !")
+		responsewriters.InternalError(rw, req, err)
+		return
+	}
+	if sets.NewString(user.GetGroups()...).Has(kuser.SystemPrivilegedGroup) {
+		// If we are the system privileged group, it might be a call from the virtual workspace
+		// in which case we also search the user in the soft impersonation header.
+		if impersonated, err := softimpersonation.UserInfoFromRequestHeader(req); err != nil {
+			responsewriters.InternalError(rw, req, err)
+			return
+		} else if impersonated != nil {
+			user = impersonated
+		}
+	}
+	requestInfo, ok := request.RequestInfoFrom(ctx)
+	if !ok {
+		responsewriters.InternalError(rw, req, errors.New("No request Info"))
+		return
+	}
+
+	if lcluster.Name == tenancyv1alpha1.RootCluster &&
+		requestInfo.IsResourceRequest &&
+		requestInfo.Verb == "get" &&
+		requestInfo.APIGroup == tenancyv1beta1.SchemeGroupVersion.Group &&
+		requestInfo.Resource == "workspaces" &&
+		requestInfo.Name == "~" {
+
+		// if we are in the special case of a `kubectl get workspace ~` request (against the 'root' workspace),
+		// we return the Home workspace definition of the current user, possibly even before its
+		// underlying ClusterWorkspace resource exists.
+
+		homeLogicalClusterName := h.getHomeLogicalClusterName(user.GetName())
+		homeClusterWorkspace, err := h.localInformers.getClusterWorkspace(homeLogicalClusterName)
+		if err != nil && !kerrors.IsNotFound(err) {
+			responsewriters.InternalError(rw, req, err)
+			return
+		}
+		if homeClusterWorkspace != nil {
+			// The home workspace for the user exists, so let's return it to the user if he requests to see its definition.
+			// We don't need to check any permission before returning the home workspace definition since,
+			// once it has been created, a home workspace is owned by the user.
+
+			homeWorkspace := &tenancyv1beta1.Workspace{}
+			projection.ProjectClusterWorkspaceToWorkspace(homeClusterWorkspace, homeWorkspace)
+			responsewriters.WriteObjectNegotiated(homeWorkspaceCodecs, negotiation.DefaultEndpointRestrictions, tenancyv1beta1.SchemeGroupVersion, rw, req, http.StatusOK, homeWorkspace)
```

Comment:
there is no authz here, correct?

Reply:
🔴

Reply:
ok, I think this is fine. It is not about workspace access (content), but only about seeing the object.